### PR TITLE
Test with coverage output, publish gcov on Azure pipelines

### PR DESCRIPTION
**Pull request by qiluo-msft**

### CODE REVIEW

**Review by kcudnik — APPROVED** [review]
This seems much simpler than this preposition: https://github.com/Azure/sonic-swss/pull/1737/files
Will this change also include coverage during vstests run or this is only for local unit tests ?

File: .azure-pipelines/build-template.yml [hunk 1]
```diff
@@ -79,6 +80,14 @@ jobs:
...
     artifact: ${{ parameters.artifact_name }}
     displayName: "Archive swss common debian packages"
+  - script: |
+      set -ex
+      # Install .NET CORE
+      curl -sSL https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -
+      sudo apt-add-repository https://packages.microsoft.com/debian/10/prod
+      sudo apt-get update
+      sudo apt-get install -y dotnet-sdk-5.0
+    displayName: "Install .NET CORE"
```
Comment (kcudnik) [inline .azure-pipelines/build-template.yml:90]: interesting, why .net is needed ?

Comment (qiluo-msft) [reply .azure-pipelines/build-template.yml:90]: Azure pipelines will report warning .Net core is missing. The task needs it to process coverage result.

**Comment by kcudnik** [comment]
should we consider disabling optimization when performing code coverage: https://gcc.gnu.org/onlinedocs/gcc/Gcov-and-Optimization.html

File: .azure-pipelines/build-template.yml [hunk 2]
```diff
@@ -53,10 +57,12 @@ jobs:
     displayName: "Install dependencies"
   - script: |
       ./autogen.sh
-      dpkg-buildpackage -us -uc -b -j$(nproc) && cp ../*.deb .
-    displayName: "Compile sonic swss common"
+      ./configure --enable-code-coverage
```
Comment (kcudnik) [inline .azure-pipelines/build-template.yml:60]: i pulled your PR locally and when i run this i get this warning:
```
config.status: config.h is unchanged
config.status: executing depfiles commands
config.status: executing libtool commands
configure: WARNING: unrecognized options: --enable-code-coverage
```

and no flags are set to coverage, but on build pipeline i can see that this param succeeded: https://dev.azure.com/mssonic/build/_build/results?buildId=26869&view=logs&j=a5b8631f-d2a8-5bf4-c5b7-ed5dd46e146c&t=93567928-e4d4-5802-f979-ee59d6645988 am i missing something?

Comment (qiluo-msft) [reply .azure-pipelines/build-template.yml:60]: Did you run above command `./autogen.sh`? I tested and believe it will fix the warning.

Comment (kcudnik) [reply .azure-pipelines/build-template.yml:60]: PS published code coverage is empty: https://dev.azure.com/mssonic/build/_build/results?buildId=26869&view=artifacts&pathAsName=false&type=publishedArtifacts

Comment (kcudnik) [reply .azure-pipelines/build-template.yml:60]: yea, i run ./autogen.sh, but it also failed, i tried manually adding ax_files, but still get error on AC_MSG_ERROR:
```
kcudnik@kcudnik-VirtualBox ~/sonic-swss-common $ ./autogen.sh
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'config'.
libtoolize: copying file 'config/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'config'.
libtoolize: copying file 'config/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
configure.ac:31: error: possibly undefined macro: AC_MSG_ERROR
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: /usr/bin/autoconf failed with exit status: 1
kcudnik@kcudnik-VirtualBox ~/sonic-swss-common $ l m4/
total 388K
drwxrwxr-x  2 kcudnik kcudnik 4,0K sie  3 12:12 .
drwxrwxr-x 11 kcudnik kcudnik 4,0K sie  3 12:12 ..
-rw-rw-r--  1 kcudnik kcudnik  914 sie  3 11:59 ax_ac_append_to_file.m4
-rw-rw-r--  1 kcudnik kcudnik  905 sie  3 12:03 ax_ac_print_to_file.m4
-rw-rw-r--  1 kcudnik kcudnik  853 sie  3 11:51 ax_add_am_macro_static.m4
-rw-rw-r--  1 kcudnik kcudnik 1,1K sie  3 12:01 ax_am_macros_static.m4
-rw-rw-r--  1 kcudnik kcudnik  12K sie  3 11:51 ax_code_coverage.m4
-rw-rw-r--  1 kcudnik kcudnik  763 sie  3 12:00 ax_file_escapes.m4
-rw-r--r--  1 kcudnik kcudnik 300K sie  3 12:12 libtool.m4
-rw-r--r--  1 kcudnik kcudnik 6,0K sie  3 12:12 lt~obsolete.m4
-rw-r--r--  1 kcudnik kcudnik  15K sie  3 12:12 ltoptions.m4
-rw-r--r--  1 kcudnik kcudnik 4,3K sie  3 12:12 ltsugar.m4
-rw-r--r--  1 kcudnik kcudnik  699 sie  3 12:12 ltversion.m4
-rw-rw-r--  1 kcudnik kcudnik 8,2K lip  7 18:45 python3.m4
```

Comment (kcudnik) [reply .azure-pipelines/build-template.yml:60]: ok, this solved the problem:
```
 sudo apt-get -y install autoconf automake libtool cmake autoconf-archive build-essential
```

**Comment by kcudnik** [comment]
> should we consider disabling optimization when performing code coverage: https://gcc.gnu.org/onlinedocs/gcc/Gcov-and-Optimization.html

seems like CODE_COVERATE override that by default:
 224 CODE_COVERAGE_CFLAGS = -O0 -g -fprofile-arcs -ftest-coverage
 225 CODE_COVERAGE_CPPFLAGS = -DNDEBUG
 226 CODE_COVERAGE_CXXFLAGS = -O0 -g -fprofile-arcs -ftest-coverage

**Comment by kcudnik** [comment]
@qiluo-msft  in your build you get the same error:
https://dev.azure.com/mssonic/build/_build/results?buildId=26869&view=logs&j=a5b8631f-d2a8-5bf4-c5b7-ed5dd46e146c&t=93567928-e4d4-5802-f979-ee59d6645988&l=313
you need to install **autoconf-archive** package additionally, and also CODE_COVERAGE_CXXFLAGS even passed at last are overrided by CXXFLAGS which by default are "-g -O2" so CODE_COVERAGE_CXXFLAGS  with "-O0" is ignored
command needed to compile this would be:
```
make CFLAGS="" CXXFLAGS=""
```
to disable optimization

**Comment by kcudnik** [comment]
maybe gcovr will need extra parameter "--exclude-throw-branches" explanation: https://github.com/gcovr/gcovr/issues/283
for out current code every SWSS_LOG_ENTER() will show untaken branch (throw) that is not taken

image
this is due to compiler generating possible exception handling in the constructor and seems like gcov is catching that as branch not executed

Also for example compiling sonic-sairedis with -O0, will also need probably swss-common compiled with -O0, since there could be artefacts, for example sizeof(std::string) is different when compiled in -O2 and -O0 since it's contain more debug info and this could lead to some weird exceptions and crashes in code

**Review by kcudnik — APPROVED** [review]

**Pull request merged**